plugins/Files/js/sagas/helpers.js

```diff
@@ -0,0 +1,161 @@
+// Helper functions for the Files sagas.
+import { List, Map } from 'immutable'
+import BigNumber from 'bignumber.js'
+import Path from 'path'
+import fs from 'fs'
+
+export const sendError = (e) => {
+	SiaAPI.showError({
+		title: 'Sia-UI Files Error',
+		content: e.toString(),
+	})
+}
+
+// siadCall: promisify Siad API calls.  Resolve the promise with `response` if the call was successful,
+// otherwise reject the promise with `err`.
+export const siadCall = (uri) => new Promise((resolve, reject) => {
+	SiaAPI.call(uri, (err, response) => {
+		if (err) {
+			reject(err)
+		} else {
+			resolve(response)
+		}
+	})
+})
+
+// Take a number of bytes and return a sane, human-readable size.
+export const readableFilesize = (bytes) => {
+	const units = ['PB', 'TB', 'GB', 'MB', 'KB', 'B']
+	let readableunit = 'B'
+	let readablesize = bytes
+	for (const unit in units) {
+		if (readablesize < 1000) {
+			readableunit = units[unit]
+			break
+		}
+		readablesize /= 1000
+	}
+	return readablesize.toFixed(2).toString() + ' ' + readableunit
+}
+
+// return a list of files filtered with path.
+// ... it's ls.
+export const ls = (files, path) => {
+	let fileList = files.filter((file) => file.uploadprogress >= 100)
+	fileList = fileList.filter((file) => file.siapath.indexOf(path) !== -1)
+	let parsedFiles = Map()
+	fileList.forEach((file) => {
+		let type = 'file'
+		const relativePath = file.siapath.substring(path.length, file.siapath.length)
+		let filename = Path.basename(relativePath)
+		let uploadprogress = Math.floor(file.uploadprogress)
+		if (relativePath.indexOf('/') !== -1) {
+			type = 'directory'
+			filename = relativePath.split('/')[0]
+			uploadprogress = ''
+		}
+		parsedFiles = parsedFiles.set(filename, {
```

Comment:
it's because `parsedFiles` should contain a list of `files` and `directories` in `path`.  Without using an `Immutable.Map()` and forEach, this function will return multiple `directories` if two or more files reside inside the same siapath folder.